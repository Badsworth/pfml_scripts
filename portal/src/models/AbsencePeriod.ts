import { LeaveReasonType } from "./LeaveReason";
import { groupBy } from "lodash";

export type AbsencePeriodRequestDecision =
  | "Approved"
  | "Cancelled"
  | "Denied"
  | "In Review"
  | "Pending"
  | "Completed"
  // Projected is (at least) used for Bonding Leave periods created automatically
  // for periods with Birth Disability as the reason qualifier
  | "Projected"
  | "Voided"
  | "Withdrawn";

export type AbsencePeriodTypes =
  | "Continuous"
  | "Intermittent"
  | "Reduced Schedule";

export class AbsencePeriod {
  absence_period_end_date: string;
  absence_period_start_date: string;
  fineos_leave_request_id: string | null = null;
  period_type: AbsencePeriodTypes;
  reason: LeaveReasonType;
  reason_qualifier_one = "";
  reason_qualifier_two = "";
  request_decision: AbsencePeriodRequestDecision;

  constructor(attrs: Partial<AbsencePeriod> = {}) {
    Object.assign(this, attrs);
  }

  static groupByReason(absence_periods: AbsencePeriod[]): {
    [reason: string]: AbsencePeriod[];
  } {
    return groupBy(absence_periods, "reason");
  }

  /**
   * @returns periods sorted newest to oldest (by start date)
   */
  static sortNewToOld(absence_periods: AbsencePeriod[]): AbsencePeriod[] {
    const periods = absence_periods.slice(); // avoids mutating the original array
    return periods.sort((a, b) => {
      return a.absence_period_start_date > b.absence_period_start_date ? -1 : 1;
    });
  }
}
