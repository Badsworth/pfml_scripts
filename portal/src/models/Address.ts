/* eslint sort-keys: ["error", "asc"] */
import BaseModel from "./BaseModel";

class Address extends BaseModel {
  // @ts-expect-error ts-migrate(2416) FIXME: Property 'defaults' in type 'Address' is not assig... Remove this comment to see the full error message
  get defaults() {
    return {
      city: null,
      line_1: null,
      line_2: null,
      state: null,
      zip: null,
    };
  }
}

export default Address;
