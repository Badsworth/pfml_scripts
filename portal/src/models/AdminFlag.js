import BaseModel from "./BaseModel";

class FlagModel extends BaseModel {
  get defaults() {
    return {
      flag_id: null,
      name: null,
      enabled: false,
      options: {},
      start: null,
      end: null,
    };
  }
}

export default FlagModel;
