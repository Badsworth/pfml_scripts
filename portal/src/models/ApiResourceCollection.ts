export interface ApiResource {
  // API resources are expected to be objects. We expect at least an ID property.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

/**
 * Readonly generic class representing a collection of a specific data model. Methods
 * return new instances of the collection rather than modifying the original collection.
 * @example const c = new ApiResourceCollection<Document>("document_id", [{ document_id: "123", name: "example" }]);
 *          c.setItem({ document_id: "456", name: "example2" });
 */
class ApiResourceCollection<TApiResource extends ApiResource> {
  idKey: keyof TApiResource;
  private map: Map<string, TApiResource>;

  constructor(idKey: keyof TApiResource, items: TApiResource[] = []) {
    this.idKey = idKey;
    this.map = new Map(
      items.map((item) => {
        return [item[idKey], item];
      })
    );
  }

  private get _mutableMap() {
    return new Map(this.map);
  }

  private _arrayFromMap(map: Map<string, TApiResource>): TApiResource[] {
    return Array.from(map.values());
  }

  get items(): TApiResource[] {
    return this._arrayFromMap(this.map);
  }

  get isEmpty(): boolean {
    return this.map.size === 0;
  }

  getItem(itemId: string): TApiResource | undefined {
    return this.map.get(itemId);
  }

  /**
   * Add or update an item in the collection. Returns a new collection with the
   * item included. Does not modify the original collection.
   */
  setItem(item: TApiResource): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap.set(item[this.idKey], item);
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Add or update items in the collection. Returns a new collection with the
   * items included. Does not modify the original collection.
   */
  setItems(items: TApiResource[] = []): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    items.forEach((item) => {
      updatedMap.set(item[this.idKey], item);
    });
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Removes an item from the collection. Returns a new collection with the item removed.
   * Does not modify the original collection.
   */
  removeItem(itemId: string): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    updatedMap.delete(itemId);

    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }
}

export default ApiResourceCollection;
