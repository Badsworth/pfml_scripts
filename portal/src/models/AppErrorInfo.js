import BaseModel from "./BaseModel";
import uniqueId from "lodash/uniqueId";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property {string} [field] - Path of the field the error is associated with
 * @property {string} key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property {string} message - Internationalized message displayed to the user (like `Error#message`)
 * @property {string} [name] - Name of the error (like `Error#name`)
 */
class AppErrorInfo extends BaseModel {
  get defaults() {
    return {
      key: uniqueId("AppErrorInfo"),
      message: null,
      name: null,
      field: null,
    };
  }
}

export default AppErrorInfo;
