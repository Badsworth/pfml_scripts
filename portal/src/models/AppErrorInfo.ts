import { uniqueId } from "lodash";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property [field] - Path of the field the error is associated with
 * @property key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property message - Internationalized message displayed to the user (like `Error#message`)
 * @property [name] - Name of the error (like `Error#name`)
 * @property [meta] - Additional error data, like the application_id associated with it
 * @property [rule] - Name of validation issue rule (e.g "min_leave_periods", "conditional", etc)
 * @property [type] - Name of validation issue type (e.g "required", "pattern", "date", etc)
 */
class AppErrorInfo {
  key: string = uniqueId("AppErrorInfo");
  message?: string | JSX.Element;
  name?: string;
  field?: string;
  meta?: Record<string, unknown>;
  rule?: string;
  type?: string;

  constructor(attrs: Partial<AppErrorInfo>) {
    Object.assign(this, attrs);
  }
}

export default AppErrorInfo;
