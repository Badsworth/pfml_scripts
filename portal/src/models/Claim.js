/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Benefits application model and enum values
 */
import BaseModel from "./BaseModel";

class Claim extends BaseModel {
  get defaults() {
    return {
      application_id: null,
      // TODO: We'll map this to the correct API field once we get into Intermittent Leave work
      avg_weekly_hours_worked: null,
      created_at: null,
      duration_type: null,
      employee_ssn: null,
      employer_benefits: [],
      employer_fein: null,
      // TODO: This field does not yet exist in the API: https://lwd.atlassian.net/browse/CP-574
      employment_status: null,
      first_name: null,
      // TODO: We'll map this to the correct API field once we get into Intermittent Leave work
      hours_off_needed: null,
      last_name: null,
      leave_details: {
        continuous_leave_periods: null,
        employer_notification_date: null,
        employer_notified: null,
        reason: null,
      },
      middle_name: null,
      other_incomes: [],
      pregnant_or_recent_birth: null,
      previous_leaves: [],
      status: null,
    };
  }
}

/**
 * Enums for the Application's `status` field
 * @enum {string}
 */
export const ClaimStatus = {
  // Submitted by the user (Step 2)
  completed: "Completed",
  // AKA: In Progress (Step 1)
  started: "Started",
  // Stored in the claims processing system (Step 3)
  submitted: "Submitted",
};

/**
 * Enums for the Application's `leave_details.employment_status` field
 * @enum {string}
 */
export const EmploymentStatus = {
  employed: "employed",
  selfEmployed: "self-employed",
  unemployed: "unemployed",
};

/**
 * Enums for the Application's `leave_details.reason` field
 * @enum {string}
 */
export const LeaveReason = {
  // TODO: We need to map some of these to the correct API fields,
  // once those enum values exist within the API
  // https://lwd.atlassian.net/browse/CP-515
  activeDutyFamily: "Care For A Family Member",
  bonding: "Child Bonding",
  medical: "Serious Health Condition - Employee",
  serviceMemberFamily: "Pregnancy/Maternity",
};

export default Claim;
