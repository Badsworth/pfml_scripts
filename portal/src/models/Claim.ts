/* eslint sort-keys: ["error", "asc"] */
import { compact, get } from "lodash";

/**
 * A record from the API's Claims table. Could be utilized by Leave Admin and Claimants.
 */
class Claim {
  absence_period_end_date: string | null = null;
  absence_period_start_date: string | null = null;
  claim_status: AbsenceCaseStatusType | null;
  claim_type_description: string | null = null;
  created_at: string;
  employee: ClaimEmployee | null;
  employer: ClaimEmployer;
  fineos_absence_id: string;
  fineos_notification_id: string;
  managed_requirements: ManagedRequirement[];

  constructor(attrs: Claim) {
    Object.assign(this, attrs);
    if (attrs.employee) {
      this.employee = new ClaimEmployee(attrs.employee);
    }
  }
}

/**
 * Employee (claimant) record associated to the Claim
 */
export class ClaimEmployee {
  first_name: string | null = null;
  last_name: string | null = null;
  middle_name: string | null = null;
  other_name: string | null = null;

  constructor(attrs: Partial<ClaimEmployee>) {
    Object.assign(this, attrs);
  }

  get fullName() {
    return compact([
      get(this, "first_name"),
      get(this, "middle_name"),
      get(this, "last_name"),
    ]).join(" ");
  }
}

/**
 * Employer record associated to the Claim
 */
export interface ClaimEmployer {
  employer_dba: string | null;
  employer_fein: string;
  employer_id: string;
}

/**
 * Managed requirements associated to the Claim
 */
export interface ManagedRequirement {
  category: string;
  created_at: string;
  follow_up_date: string | null;
  responded_at: string | null;
  status: "Open" | "Complete" | "Suppressed";
  type: string;
}

/**
 * Enums for `claim.claim_status` field
 * This is the subset of potential values returned by the API that we show to the user.
 * We currently ignore other potential values like "Adjudication", "Intake In Progress", "Unknown"
 * @enum {string}
 */
export const AbsenceCaseStatus = {
  approved: "Approved",
  closed: "Closed",
  completed: "Completed",
  declined: "Declined",
} as const;

export type AbsenceCaseStatusType =
  typeof AbsenceCaseStatus[keyof typeof AbsenceCaseStatus];

export default Claim;
