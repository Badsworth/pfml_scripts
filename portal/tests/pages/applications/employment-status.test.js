import {
  MockClaimBuilder,
  renderWithAppLogic,
  simulateEvents,
} from "../../test-utils";
import { EmploymentStatus } from "../../../src/models/Claim";
import EmploymentStatusPage from "../../../src/pages/applications/employment-status";

jest.mock("../../../src/hooks/useAppLogic");

const setup = (claimAttrs) => {
  const { appLogic, claim, wrapper } = renderWithAppLogic(
    EmploymentStatusPage,
    {
      claimAttrs: claimAttrs || new MockClaimBuilder().create(),
    }
  );
  const { changeField, changeRadioGroup, submitForm } = simulateEvents(wrapper);
  return {
    appLogic,
    changeField,
    changeRadioGroup,
    claim,
    submitForm,
    wrapper,
  };
};

const notificationFeinQuestionWrapper = (wrapper) => {
  return wrapper.find({ name: "employer_fein" });
};

describe("EmploymentStatusPage", () => {
  describe("when claimantShowEmploymentStatus feature flag is disabled", () => {
    beforeEach(() => {
      process.env.featureFlags = {
        claimantShowEmploymentStatus: false,
      };
    });

    it("renders the page without the employment status field", () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find("Trans").dive()).toMatchSnapshot();
    });

    it("submits status and FEIN", () => {
      const { appLogic, claim, changeField, submitForm } = setup();
      const testFein = 123456789;
      changeField("employer_fein", testFein);

      submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(
        claim.application_id,
        {
          employment_status: EmploymentStatus.employed,
          employer_fein: testFein,
        }
      );
    });
  });

  describe("when claimantShowEmploymentStatus feature flag is enabled", () => {
    beforeEach(() => {
      process.env.featureFlags = {
        claimantShowEmploymentStatus: true,
      };
    });

    it("renders the page with the employment status field", () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });

    describe("when user selects employed in MA as their employment status", () => {
      it("shows FEIN question", () => {
        const { changeRadioGroup, wrapper } = setup();
        changeRadioGroup("employment_status", EmploymentStatus.employed);

        expect(
          notificationFeinQuestionWrapper(wrapper)
            .parents("ConditionalContent")
            .prop("visible")
        ).toBeTruthy();
      });

      it("submits status and FEIN", () => {
        const {
          appLogic,
          claim,
          changeField,
          changeRadioGroup,
          submitForm,
        } = setup();
        changeRadioGroup("employment_status", EmploymentStatus.employed);
        const testFein = 123456789;
        changeField("employer_fein", testFein);
        submitForm();

        expect(appLogic.claims.update).toHaveBeenCalledWith(
          claim.application_id,
          {
            employment_status: EmploymentStatus.employed,
            employer_fein: testFein,
          }
        );
      });
    });

    describe("when user selects self-employed as their employment status", () => {
      it("hides FEIN question", () => {
        const { changeRadioGroup, wrapper } = setup();
        changeRadioGroup("employment_status", EmploymentStatus.selfEmployed);
        expect(
          notificationFeinQuestionWrapper(wrapper)
            .parents("ConditionalContent")
            .prop("visible")
        ).toBeFalsy();
      });

      it("submits status and empty FEIN", () => {
        const { appLogic, claim, changeRadioGroup, submitForm } = setup();
        changeRadioGroup("employment_status", EmploymentStatus.selfEmployed);
        submitForm();

        expect(appLogic.claims.update).toHaveBeenCalledWith(
          claim.application_id,
          {
            employment_status: EmploymentStatus.selfEmployed,
            employer_fein: null,
          }
        );
      });
    });

    describe("when user selects unemployed as their employment status", () => {
      it("hides FEIN question", () => {
        const { changeRadioGroup, wrapper } = setup();
        changeRadioGroup("employment_status", EmploymentStatus.unemployed);
        expect(
          notificationFeinQuestionWrapper(wrapper)
            .parents("ConditionalContent")
            .prop("visible")
        ).toBeFalsy();
      });
    });

    describe("when claim has existing employment status", () => {
      it("submits status and FEIN without changing fields", () => {
        const test = new MockClaimBuilder().employed().create();
        const { appLogic, claim, submitForm } = setup(test);

        submitForm();

        expect(appLogic.claims.update).toHaveBeenCalledWith(
          claim.application_id,
          {
            employment_status: EmploymentStatus.employed,
            employer_fein: "12-3456789",
          }
        );
      });
    });
  });
});
