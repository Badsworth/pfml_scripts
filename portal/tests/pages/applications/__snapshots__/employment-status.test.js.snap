// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmploymentStatusPage when claimantShowEmploymentStatus feature flag is disabled renders the page without the employment status field 1`] = `
<div>
  <form
    class="usa-form"
    method="post"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-2 font-heading-sm line-height-sans-3"
      tabindex="-1"
    >
      Employment information
    </h1>
    <div
      class="usa-alert usa-alert--info c-alert--neutral"
      role="region"
      tabindex="-1"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          <p>
            If you’re self-employed or unemployed, apply by calling the Department of Family and Medical Leave Contact Center at 
            <a
              href="tel:(833) 344-7365"
            >
              (833) 344‑7365
            </a>
            .
          </p>
          <p>
            You can apply online if you’re currently employed in Massachusetts.
          </p>
        </div>
      </div>
    </div>
    <div
      class="usa-form-group"
    >
      <label
        class="usa-label text-bold font-heading-lg line-height-sans-3 margin-bottom-1 maxw-tablet"
        for="InputText1"
        id="InputText1_label"
      >
        What is your employer’s Employer Identification Number (EIN)?
      </label>
      <span
        class="display-block line-height-sans-5 measure-5 usa-intro"
        id="InputText1_hint"
      >
        This number is 9 digits. You can find this number on all notices your employer sent about Paid Family and Medical Leave. You can also find it on your W‑2 or 1099‑MISC. Ask your employer if you need help getting this information.
      </span>
      <div
        class=""
      >
        <input
          aria-labelledby="InputText1_label InputText1_hint InputText1_error"
          class="usa-input maxw-mobile-lg"
          id="InputText1"
          inputmode="numeric"
          name="employer_fein"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="usa-button position-relative margin-top-4"
      type="submit"
    >
      Save and continue
    </button>
  </form>
</div>
`;

exports[`EmploymentStatusPage when claimantShowEmploymentStatus feature flag is enabled renders the page with the employment status field 1`] = `
<div>
  <form
    class="usa-form"
    method="post"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-2 font-heading-sm line-height-sans-3"
      tabindex="-1"
    >
      Employment information
    </h1>
    <fieldset
      class="usa-fieldset margin-top-3 usa-form-group"
    >
      <legend
        class="usa-label text-bold usa-legend font-heading-lg line-height-sans-3 margin-bottom-1 maxw-tablet"
      >
        What is your employment status?
      </legend>
      <span
        class="display-block line-height-sans-5 measure-5 usa-intro"
      >
        <details
          class="c-details"
        >
          <summary
            class="margin-bottom-2 text-primary text-underline font-ui-xs"
          >
            What if I’ve been furloughed?
          </summary>
          <div
            class="margin-bottom-2"
          >
            If your hours have been cut or significantly reduced but you have not been laid off, select "I’m employed in Massachusetts"
          </div>
        </details>
      </span>
      <div
        class="margin-top-3"
      >
        <div
          class="usa-radio measure-5 bg-transparent"
        >
          <input
            class="usa-radio__input"
            id="InputChoice13"
            name="employment_status"
            type="radio"
            value="Employed"
          />
          <label
            class="usa-radio__label"
            for="InputChoice13"
          >
            I’m employed in Massachusetts
          </label>
        </div>
        <div
          class="usa-radio measure-5 bg-transparent"
        >
          <input
            class="usa-radio__input"
            id="InputChoice14"
            name="employment_status"
            type="radio"
            value="Unemployed"
          />
          <label
            class="usa-radio__label"
            for="InputChoice14"
          >
            I’m unemployed
          </label>
        </div>
        <div
          class="usa-radio measure-5 bg-transparent"
        >
          <input
            class="usa-radio__input"
            id="InputChoice15"
            name="employment_status"
            type="radio"
            value="Self-Employed"
          />
          <label
            class="usa-radio__label"
            for="InputChoice15"
          >
            I’m self-employed
          </label>
        </div>
      </div>
    </fieldset>
    <button
      class="usa-button position-relative margin-top-4"
      type="submit"
    >
      Save and continue
    </button>
  </form>
</div>
`;
